src/repository.rs

```diff
@@ -117,30 +121,85 @@ where
     T: RepositoryStorage<D>,
     D: DataInterchange + Sync,
 {
-    fn store_metadata<'a, R>(
+    fn store_metadata<'a>(
         &'a self,
         meta_path: &'a MetadataPath,
         version: &'a MetadataVersion,
-        metadata: R,
-    ) -> BoxFuture<'a, Result<()>>
-    where
-        R: AsyncRead + Send + Unpin + 'a,
-    {
+        metadata: &'a mut (dyn AsyncRead + Send + Unpin + 'a),
+    ) -> BoxFuture<'a, Result<()>> {
         (**self).store_metadata(meta_path, version, metadata)
     }
 
-    fn store_target<'a, R>(
+    fn store_target<'a>(
         &'a self,
-        target: R,
+        target: &'a mut (dyn AsyncRead + Send + Unpin + 'a),
         target_path: &'a TargetPath,
-    ) -> BoxFuture<'a, Result<()>>
-    where
-        R: AsyncRead + Send + Unpin + 'a,
-    {
+    ) -> BoxFuture<'a, Result<()>> {
+        (**self).store_target(target, target_path)
+    }
+}
+
+impl<T, D> RepositoryStorageProvider<D> for &T
+where
+    T: RepositoryStorage<D> + RepositoryProvider<D>,
+    D: DataInterchange + Sync,
+{
+}
+
+impl<T, D> RepositoryStorage<D> for Box<T>
+where
+    T: RepositoryStorage<D> + ?Sized,
+    D: DataInterchange + Sync,
+{
+    fn store_metadata<'a>(
+        &'a self,
+        meta_path: &'a MetadataPath,
+        version: &'a MetadataVersion,
+        metadata: &'a mut (dyn AsyncRead + Send + Unpin + 'a),
+    ) -> BoxFuture<'a, Result<()>> {
+        (**self).store_metadata(meta_path, version, metadata)
+    }
+
+    fn store_target<'a>(
+        &'a self,
+        target: &'a mut (dyn AsyncRead + Send + Unpin + 'a),
+        target_path: &'a TargetPath,
+    ) -> BoxFuture<'a, Result<()>> {
         (**self).store_target(target, target_path)
     }
 }
 
+impl<T, D> RepositoryProvider<D> for Box<T>
+where
+    T: RepositoryProvider<D> + ?Sized,
+    D: DataInterchange + Sync,
+{
+    fn fetch_metadata<'a>(
+        &'a self,
+        meta_path: &'a MetadataPath,
+        version: &'a MetadataVersion,
+        max_length: Option<usize>,
+        hash_data: Option<(&'static HashAlgorithm, HashValue)>,
+    ) -> BoxFuture<'a, Result<Box<dyn AsyncRead + Send + Unpin>>> {
+        (**self).fetch_metadata(meta_path, version, max_length, hash_data)
+    }
+
+    fn fetch_target<'a>(
+        &'a self,
+        target_path: &'a TargetPath,
+        target_description: &'a TargetDescription,
+    ) -> BoxFuture<'a, Result<Box<dyn AsyncRead + Send + Unpin>>> {
+        (**self).fetch_target(target_path, target_description)
+    }
+}
+
+impl<T, D> RepositoryStorageProvider<D> for Box<T>
```

Comment:
I wonder if we could define a blanket impl for all T: Storage + Provider instead of doing manual impls for each concrete type

Reply:
done